torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -221,11 +332,38 @@ def __init__(
 
         self._optim_defaults = defaults
         self._optim_constructor = optimizer_class
-        self._init_local_optimizer()
-
+        functional_optims = ZeroRedundancyOptimizer.functional_optims
+        self._is_functional_optim = optimizer_class in functional_optims
+
+        if overlap_with_ddp and not self._is_functional_optim:
+            error_msg = "Overlapping with DDP requires a functional optimizer"
+            functional_optim_map = ZeroRedundancyOptimizer.functional_optim_map
+            if optimizer_class in functional_optim_map:
+                functional_optim_name = functional_optim_map[optimizer_class].__name__
+                raise ValueError(
+                    error_msg + f"; use {functional_optim_name} instead"
```

Comment:
This looks fine for now. But let's add a followup PR to automatically do the translation for users, similar to how `DistributedOptimizer` handles this.